sql: support `nearest_only` argument for with_min_timestamp/max_staleness

otan opened the pull request:
Release note (sql change): Introduced a `nearest_only` argument for
`with_min_timestamp`/`with_max_staleness`, which enforces bounded
staleness reads only talk to the nearest replica.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/67837)
<!-- Reviewable:end -->

nvb reviewed (commented):
Thanks!

I'm a little concerned with the ambiguity of the term "local", both because it seems imprecise ("local to what?") and because it won't match the semantics we're planning on giving the flag. I imagine that "local" is meant to mean "with the same region locality tier as the gateway". So that would imply that if no replicas for a range are in the same region as the gateway, then we should always fail. Similarly, it would imply that if two replicas for a range are in the same region as the gateway, then we should try both but not try any other replicas.

That's not quite what we are planning to implement. Instead, what we'll implement will be more aligned with "nearest_only" or "closest_only". We'll use the closest available replica (like the info text mentions) regardless of whether it is in the same region as the gateway, and that's the only replica we'll try to use.

> Reviewed 5 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67837#-:-Mf5wOTi7Hq_4sclzxWF:b-h2o58d)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[docs/RFCS/20210519_bounded_staleness_reads.md, line 94 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67837#-Mf5v1HGDywBecN43RwA:-Mf5v1HGDywBecN43RwB:b-hp1afm) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5209721e106549360f5af54c428101ce8562ed6/docs/RFCS/20210519_bounded_staleness_reads.md#L94)):*
> ```
> `local_only`argument  that
> ```

nit: the spacing looks off here. Same thing with down below.
___
*[docs/RFCS/20210519_bounded_staleness_reads.md, line 175 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67837#-Mf5v8oJ3f6CphrGP0Yt:-Mf5v8oJ3f6CphrGP0Yu:bgfjan0) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5209721e106549360f5af54c428101ce8562ed6/docs/RFCS/20210519_bounded_staleness_reads.md#L175)):*
> ```
> if the table descriptor is not available locally
> ```

What do you mean by this? Do you mean that we will reject the query if the `system.descriptor` table's leaseholder is not local and we don't have a lease on the table descriptor at the time of the query?
___
*[docs/RFCS/20210519_bounded_staleness_reads.md, line 540 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67837#-Mf5veSt98f7QeFkI5DD:-Mf5veSt98f7QeFkI5DE:bskhsx1) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5209721e106549360f5af54c428101ce8562ed6/docs/RFCS/20210519_bounded_staleness_reads.md#L540)):*
> ```
> This is still the case if `local_only` is specified.
> ```

Why does `local_only` have an impact on this? We'll still iterate up the schema history when `local_only` is set and a newer version is unavailable. It's only when we get to the earliest usable schema version that `local_only` makes a difference.
___
*[pkg/sql/sem/builtins/builtins.go, line 2487 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67837#-Mf5w5r02uILqSJt5FFS:-Mf5w5r02uILqSJt5FFT:b-pb15y4) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5209721e106549360f5af54c428101ce8562ed6/pkg/sql/sem/builtins/builtins.go#L2487)):*
> ```Go
> 				return withMinTimestamp(ctx, tree.MustBeDTimestampTZ(args[0]).Time)
> 			},
> 			PreferredOverload: true,
> ```

Why do both `WithMinTimestampFunctionName` functions set `PreferredOverload` to true but both `WithMaxStalenessFunctionName` leave it as false?


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
i've changed it to nearest_only.  i misunderstood it to mean "local" as in local to the region, but the difference you describe makes sense.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67837#-:-Mf64WGY6TqnYle6m-Ce:byozdny)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[docs/RFCS/20210519_bounded_staleness_reads.md, line 175 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67837#-Mf5v8oJ3f6CphrGP0Yt:-Mf62Hbk22EsR6bz7sBS:b-5sqmyh) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5209721e106549360f5af54c428101ce8562ed6/docs/RFCS/20210519_bounded_staleness_reads.md#L175)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

> ```
> if the table descriptor is not available locally
> ```

What do you mean by this? Do you mean that we will reject the query if the `system.descriptor` table's leaseholder is not local and we don't have a lease on the table descriptor at the time of the query?
</blockquote></details>

i've clarified this a little bit.
___
*[docs/RFCS/20210519_bounded_staleness_reads.md, line 540 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67837#-Mf5veSt98f7QeFkI5DD:-Mf63gSUEu0tq8zCVuyI:b-6rnbt2) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5209721e106549360f5af54c428101ce8562ed6/docs/RFCS/20210519_bounded_staleness_reads.md#L540)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

> ```
> This is still the case if `local_only` is specified.
> ```

Why does `local_only` have an impact on this? We'll still iterate up the schema history when `local_only` is set and a newer version is unavailable. It's only when we get to the earliest usable schema version that `local_only` makes a difference.
</blockquote></details>

fixed this up a bit
___
*[pkg/sql/sem/builtins/builtins.go, line 2487 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67837#-Mf5w5r02uILqSJt5FFS:-Mf61EIuEGalxBNXV92G:b-r63nlo) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5209721e106549360f5af54c428101ce8562ed6/pkg/sql/sem/builtins/builtins.go#L2487)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Why do both `WithMinTimestampFunctionName` functions set `PreferredOverload` to true but both `WithMaxStalenessFunctionName` leave it as false?
</blockquote></details>

oops, that doesn't need to be set.
i've removed PreferredOverload mentions.


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 4 of 5 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67837)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten and @otan)

___
*[docs/RFCS/20210519_bounded_staleness_reads.md, line 6 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67837#-Mf69L_zEzicIuW_ORnP:-Mf69La-1jGOmFrwtjby:bto7xkz) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5209721e106549360f5af54c428101ce8562ed6/docs/RFCS/20210519_bounded_staleness_reads.md#L6)):*
> ```Markdown
> - Authors: Nathan VanBenschoten, Rebecca Taft, Oliver Tan
> - RFC PR: #66020.
> - Cockroach Issue: #25405.
> ```

nit: while here, want to delete the period above as well?
___
*[docs/RFCS/20210519_bounded_staleness_reads.md, line 175 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67837#-Mf5v8oJ3f6CphrGP0Yt:-Mf69rr0CrpyhEHEqdX0:bq4b8ts) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5209721e106549360f5af54c428101ce8562ed6/docs/RFCS/20210519_bounded_staleness_reads.md#L175)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

i've clarified this a little bit.
</blockquote></details>

I see. Is this actually what we are planning on implementing though? I wasn't thinking that `nearest_only` would have any interaction with table leases. It seems like a pretty rough failure mode that if a table descriptor is not leased at some point in time and we only have bounded staleness reads on the table, we'll never grab a lease and begin rejecting all queries. 

We assume that the `system.descriptor` table is always going to be available, so a missing lease will lead to a blip in latency, but I think that's ok since it won't block indefinitely during a region outage. If users truly want a hard latency cap for all operations, they should still set timeouts.

Put another way, I don't think we need to make the `nearest_only` flag result in all cross-region operations being rejected. I think it's enough to keep it very constrained to its interaction with dynamic timestamp selection.
___
*[docs/RFCS/20210519_bounded_staleness_reads.md, line 12 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/67837#-Mf69VKB2UPyrnxioH3G:-Mf69VKB2UPyrnxioH3H:bwgkyv3) ([raw file](https://github.com/cockroachdb/cockroach/blob/c1b895a4a522fdbb585be14a149f31dd09f8756b/docs/RFCS/20210519_bounded_staleness_reads.md#L12)):*
> ```Markdown
> Bounded staleness reads are a form of historical read-only queries that use a
> dynamic, system-determined timestamp, subject to a user-provided staleness
> bound, to read nearby local replicas while minimizing data staleness. They
> ```

Should we revert this change?
___
*[docs/RFCS/20210519_bounded_staleness_reads.md, line 39 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/67837#-Mf69fHT6i1K7T1jK19M:-Mf69fHT6i1K7T1jK19N:b-twj3un) ([raw file](https://github.com/cockroachdb/cockroach/blob/c1b895a4a522fdbb585be14a149f31dd09f8756b/docs/RFCS/20210519_bounded_staleness_reads.md#L39)):*
> ```
> but avoids the use of the leaseholder.
> ```
It doesn't even necessarily do this. The leaseholder can be used if it is the closest replica.
___
*[pkg/sql/sem/builtins/builtins.go, line 7649 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/67837#-Mf6C3pGB0jcI8RVFnUq:-Mf6C3pGB0jcI8RVFnUr:b-3fot6p) ([raw file](https://github.com/cockroachdb/cockroach/blob/c1b895a4a522fdbb585be14a149f31dd09f8756b/pkg/sql/sem/builtins/builtins.go#L7649)):*
> ```Go
> const nearestOnlyInfo = `
> 
> If nearest_only is set to true, reads that cannot be served using a nearby replica
> ```

`s/a nearby replica/the nearest available replica/`


<!-- Sent from Reviewable.io -->

otan committed:
sql: support `nearest_only` argument for with_min_timestamp/max_staleness

Release note (sql change): Introduced a `nearest_only` argument for
`with_min_timestamp`/`with_max_staleness`, which enforces bounded
staleness reads only talk to the nearest replica.

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67837)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[docs/RFCS/20210519_bounded_staleness_reads.md, line 175 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67837#-Mf5v8oJ3f6CphrGP0Yt:-Mf6E1EP1_WrbhMli9rm:be5i5m3) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5209721e106549360f5af54c428101ce8562ed6/docs/RFCS/20210519_bounded_staleness_reads.md#L175)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I see. Is this actually what we are planning on implementing though? I wasn't thinking that `nearest_only` would have any interaction with table leases. It seems like a pretty rough failure mode that if a table descriptor is not leased at some point in time and we only have bounded staleness reads on the table, we'll never grab a lease and begin rejecting all queries. 

We assume that the `system.descriptor` table is always going to be available, so a missing lease will lead to a blip in latency, but I think that's ok since it won't block indefinitely during a region outage. If users truly want a hard latency cap for all operations, they should still set timeouts.

Put another way, I don't think we need to make the `nearest_only` flag result in all cross-region operations being rejected. I think it's enough to keep it very constrained to its interaction with dynamic timestamp selection.
</blockquote></details>

i see. so system.descriptor is always available on some replica is the assumption, i think that's reasonable. i was confusing it with my initial interpretation of local.
___
*[docs/RFCS/20210519_bounded_staleness_reads.md, line 12 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/67837#-Mf69VKB2UPyrnxioH3G:-Mf6Daa3A2pUvpziBtpV:bb7m3jc) ([raw file](https://github.com/cockroachdb/cockroach/blob/c1b895a4a522fdbb585be14a149f31dd09f8756b/docs/RFCS/20210519_bounded_staleness_reads.md#L12)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should we revert this change?
</blockquote></details>

oops


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 3 of 3 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67837#-:-Mf6JaKJDtIAjgDwEx1t:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan)



<!-- Sent from Reviewable.io -->

otan commented:
bors r=nvanbenschoten
